### PR TITLE
Use scheduleOrThrow in MergeTree data selection and initialization to avoid deadlocks

### PR DESCRIPTION
**Pull request by seandhaynes**
At the moment, the use of `scheduleOrThrowOnError` means we don't currently have a timeout when waiting for threads to be allocated. So if you reach a point where you saturate all threads available in the global pool whilst in a loop or similar, ClickHouse can enter a deadlocked state as mentioned on https://github.com/ClickHouse/ClickHouse/issues/51820

Using `scheduleOrThrow` means rather than ClickHouse getting stuck/deadlocked, we get errors such as this in query threads:
```
2024.01.24 16:01:16.251258 [ 430 ] {787da7df-8198-414b-a367-747c96f053d3} <Error> DynamicQueryHandler: Code: 439. DB::Exception: Cannot schedule a task: cannot allocate thread 
```

Co-authored with @nikitamikhaylov - Thank you for the work on this once again, @nikitamikhaylov !

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Avoid possible deadlock during MergeTree index analysis when scheduling threads in a saturated service.

Use scheduleOrThrow in MergeTree data selection and initialization to avoid deadlocks

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit b63b381d5866f877ac7ee0e54efbb77256ef3a53 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/59427/b63b381d5866f877ac7ee0e54efbb77256ef3a53/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/59427/b63b381d5866f877ac7ee0e54efbb77256ef3a53/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Comment by seandhaynes** [comment]
Hm, this is related: https://s3.amazonaws.com/clickhouse-test-reports/59427/a93fb83cf618e1199527c9a8e59e67e6e69d8fd1/bugfix_validate_check.html

Weird, will take a look.

File: src/Databases/DatabaseOnDisk.cpp [hunk 1]
```diff
@@ -644,19 +646,24 @@ void DatabaseOnDisk::iterateMetadataFiles(ContextPtr local_context, const Iterat
             throw Exception(ErrorCodes::INCORRECT_FILE_NAME, "Incorrect file extension: {} in metadata directory {}", file_name, getMetadataPath());
     }
 
+    /// Don't wait forever for threads from the Global thread pool. Otherwise, ClickHouse can enter a deadlocked state
+    /// where it blocks infinitely on startup waiting for threads.
+    const std::chrono::microseconds lock_acquire_timeout_microseconds = std::chrono::seconds{getContext()->getSettingsRef().lock_acquire_timeout};
```
Comment (CheSema) [inline src/Databases/DatabaseOnDisk.cpp:651]: lets just do
`getContext()->getSettingsRef().lock_acquire_timeout.totalMicroseconds();`

it is a bit tangle how it is converted to std::chrono. `getContext()->getSettingsRef().lock_acquire_timeout` has a type as `SettingFieldTimespan`.

File: src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp [hunk 2]
```diff
@@ -746,8 +747,17 @@ RangesInDataParts MergeTreeDataSelectExecutor::filterPartsByPrimaryKeyAndSkipInd
                 CurrentMetrics::MergeTreeDataSelectExecutorThreadsScheduled,
                 num_threads);
 
+            SCOPE_EXIT({ pool.wait(); });
```
Comment (CheSema) [inline src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp:750]: Here is an interesting case could be.

`pool.wait();` -- can throw an exception from other failed task. You dont what it when you already got some from `pool.scheduleOrThrow()`.

Just leave that as it was. Than d-tor ThreadPool will wait and mute other exception.

**Review by CheSema — CHANGES_REQUESTED** [review]
The comments above.

**Comment by CheSema** [comment]
Any updates?

**Comment by seandhaynes** [comment]
Thanks for the nudge, @CheSema  and sorry for the delay. I'm addressing your feedback today and will update the PR shortly.

**Review by alexey-milovidov — APPROVED** [review]

**Review by alexey-milovidov — APPROVED** [review]

**Comment by seandhaynes** [comment]
Thank you for dealing with this, @alexey-milovidov ! Sorry for dropping off it.

**Pull request merged**